### PR TITLE
Sfraczek/conv bn fuse pass

### PR DESCRIPTION
**Pull request by sfraczek**
@Superjomn, I was wondering if you could suggest what patterns should be added to graph_pattern_detector.* and what should be moved to conv_bn_fuse_pass.* files?

### CODE REVIEW

**Comment by Superjomn** [comment]
The general patterns are better to move to `graph_pattern_detector.h::patterns` for reuse, while the rare patterns are better to leave in `conv_bn_fuse_pass`

**Comment by sfraczek** [comment]
@luotao1 Why did my build fail? Is it because of this:
```
09:51:32][Step 1/1] The following tests FAILED:
[09:51:32][Step 1/1] 	352 - test_parallel_executor_test_while_train (Failed)
[09:51:32][Step 1/1] Errors while running CTest
[09:51:46][Step 1/1] Process exited with code 8
[09:51:46][Step 1/1] Process exited with code 8
[09:51:46][Step 1/1] Step Build and test (Command Line) failed
```
? 

The second one failed because it ran out of available storage on hard drive.

**Comment by luotao1** [comment]
The ci machine encounters some problems, you can push again.

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+
+std::unique_ptr<ir::Graph> ConvBNFusePass::ApplyImpl(
+    std::unique_ptr<ir::Graph> graph) const {
+  PADDLE_ENFORCE(graph.get());
+  FusePassBase::Init(name_scope_, graph.get());
+
+  auto* scope = param_scope();
+  PADDLE_ENFORCE(scope);
+
+  std::unordered_set<Node*> nodes2delete;
```
Comment (kbinias) [inline paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: Unused variable

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+namespace paddle {
+namespace framework {
+namespace ir {
+
+LoDTensor tensor_apply(LoDTensor* vec, float (*f)(float)) {
+  LoDTensor vec_y;
+  vec_y.Resize(vec->dims());
+  float* x = vec->mutable_data<float>(platform::CPUPlace());
+  float* y = vec_y.mutable_data<float>(platform::CPUPlace());
+  for (int i = 0; i < vec->numel(); i++) {
```
Comment (kbinias) [inline paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: Consider size_t

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+    y[i] = f(x[i]);
+  }
+  return vec_y;
+}
+
+void tensor_apply_inplace(LoDTensor* vec, float (*f)(float)) {
+  float* data = vec->mutable_data<float>(platform::CPUPlace());
+  for (int i = 0; i < vec->numel(); i++) {
```
Comment (kbinias) [inline paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: Consider size_t

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+template <typename BinaryOperation>
+LoDTensor tensor_apply_eltwise(LoDTensor vec_a, LoDTensor vec_b,
+                               BinaryOperation f) {
+  PADDLE_ENFORCE_EQ(vec_a.dims(), vec_b.dims());
+  LoDTensor vec_y;
+  vec_y.Resize(vec_a.dims());
+  float* a = vec_a.mutable_data<float>(platform::CPUPlace());
+  float* b = vec_b.mutable_data<float>(platform::CPUPlace());
+  float* y = vec_y.mutable_data<float>(platform::CPUPlace());
+  for (int i = 0; i < vec_a.numel(); i++) {
```
Comment (kbinias) [inline paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: Consider size_t

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+                                         BinaryOperation f) {
+  PADDLE_ENFORCE_GT(vec_a.dims().size(), vec_b.dims().size());
+  PADDLE_ENFORCE_EQ(vec_a.dims()[0], vec_b.dims()[0]);
+  LoDTensor vec_y;
+  vec_y.Resize(vec_a.dims());
+  float* a = vec_a.mutable_data<float>(platform::CPUPlace());
+  float* b = vec_b.mutable_data<float>(platform::CPUPlace());
+  float* y = vec_y.mutable_data<float>(platform::CPUPlace());
+  size_t b_size = vec_b.dims()[0];
+  for (int i = 0; i < vec_a.numel(); ++i) {
```
Comment (kbinias) [inline paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: Consider size_t in all iterators

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+
+std::unique_ptr<ir::Graph> ConvEltwiseAddBNFusePass::ApplyImpl(
+    std::unique_ptr<ir::Graph> graph) const {
+  PADDLE_ENFORCE(graph.get());
+  FusePassBase::Init(name_scope_, graph.get());
+
+  auto* scope = param_scope();
+  PADDLE_ENFORCE(scope);
+
+  std::unordered_set<Node*> nodes2delete;
```
Comment (kbinias) [inline paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: Unused variable

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+    auto* mean_tensor =
+        scope->FindVar(bn_mean->Name())->GetMutable<LoDTensor>();
+
+    auto std_tensor = LoDTensor();
+    std_tensor.Resize(bn_bias_tensor->dims());
+    std_tensor =
+        tensor_apply(variance_tensor, [](float x) { return x + 1e-5f; });
+
+    tensor_apply_inplace(&std_tensor, std::sqrt);
+    auto tmp_tensor =
```
Comment (kbinias) [inline paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: Consider increasing readability of the code, e.g.:
{
    auto tensor_minus = 
        tensor_apply_eltwise(*bias_y_tensor, *mean_tensor, std::minus<float>());
    auto tensor_multi = 
        tensor_apply_eltwise(tensor_minus , tmp_tensor, std::multiplies<float>());
    *bias_y_tensor = 
        tensor_apply_eltwise(tensor_multi, *bn_bias_tensor, std::plus<float>());
}

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+        tensor_apply_eltwise(*scale_tensor, std_tensor, std::divides<float>());
+    *bias_y_tensor = tensor_apply_eltwise(
```
Comment (kbinias) [inline paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: Consider increasing readability of the code

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc (hunk 1 above)
Comment (sfraczek) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: ok

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc (hunk 2 above)
Comment (sfraczek) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: the return value is actually int64_t so maybe I will use it?

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc (hunk 3 above)
Comment (sfraczek) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: as above

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc (hunk 4 above)
Comment (sfraczek) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: as above

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc (hunk 5 above)
Comment (sfraczek) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: as above

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc (hunk 6 above)
Comment (sfraczek) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: ok

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc (hunk 7 above)
Comment (sfraczek) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: ok

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc (hunk 8 above)
Comment (sfraczek) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: as above

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+    GET_IR_NODE_FROM_SUBGRAPH(bn_saved_mean, bn_saved_mean, conv_bn_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(bn_saved_variance, bn_saved_variance,
+                              conv_bn_pattern);
+
+    // Create y (bias) variable
+    VarDesc y_desc(patterns::PDNodeName(name_scope_, "eltwise_y_in"));
+    auto* y_var_node = g->CreateVarNode(&y_desc);
+    auto* bias_y_tensor =
+        scope->Var(y_var_node->Name())->GetMutable<LoDTensor>();
+    std::fill_n(bias_y_tensor->mutable_data<float>(platform::CPUPlace()),
```
Comment (hshen14) [inline paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: We don't need to fill in the data with initialized value, since it will be re-initialized later.

Comment (sfraczek) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: Do you mean that I wrote fill_n twice?

**Comment by luotao1** [comment]
please see https://github.com/PaddlePaddle/Paddle/pull/13553#issuecomment-424683499 to run the TeamCity

**Comment by sfraczek** [comment]
@luotao1 TeamCity says "no space left on device" again.

**Review by luotao1 — COMMENTED** [review]
I test this PASS with #13663. In mkl mode, there is diff before and after `conv+bn` fuse pass. 
```
97: /home/luotao/Paddle/paddle/fluid/inference/tests/api/tester_helper.h:68: Failure
97: The difference between pdata_ref[j] and pdata[j] is 0.00103759765625, which exceeds 1e-3, where
97: pdata_ref[j] evaluates to -264.8758544921875,
97: pdata[j] evaluates to -264.87481689453125, and
97: 1e-3 evaluates to 0.001.
```
@sfraczek You can have a test for it.

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+    IR_NODE_LINK_TO(bias, bn_out);
+
+    found_conv_bn_count++;
+  };
+
+  gpd(graph.get(), handler);
+
+  AddStatis(found_conv_bn_count);
+  return graph;
+}
```
Comment (luotao1) [inline paddle/fluid/framework/ir/conv_bn_fuse_pass.cc:319]: `ConvEltwiseAddBNFusePass` and `ConvBNFusePass` have many similar codes except `Get y (bias) variable`. Could you abstract a common function to simplify these two functions?

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+    // Find bias from BN
+    auto* bn_bias_var = scope->FindVar(bn_bias->Name());
+    auto* bn_bias_tensor = bn_bias_var->GetMutable<LoDTensor>();
+
+    // Get y (bias) variable
+    auto* bias_y_tensor =
+        scope->FindVar(bias_y->Name())->GetMutable<LoDTensor>();
+
+    PADDLE_ENFORCE_EQ(bias_y_tensor->dims(), bn_bias_tensor->dims());
+
```
Comment (luotao1) [inline paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: It's better to add some formulas here, like

https://github.com/PaddlePaddle/Paddle/blob/35b713c3fd63e98380efa0dd1848b9f9a8223af8/python/paddle/fluid/transpiler/inference_transpiler.py#L247-L258

Comment (sfraczek) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: I'm not sure where to add a block comment like this.

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc (hunk 10 above)
Comment (sfraczek) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc:319]: Ok.

**Comment by sfraczek** [comment]
Hi @luotao1,
my colleagues are working on a fix for `elementwise_add_mkldnn_op`. This operator might compute bad numbers for some specific input dimensions (the operators created by `conv_bn_fuse` pass). Until it is done, the output after running `conv_bn_fuse` (with FLAGS_use_mkldnn=1) without following it with `conv_bias_mkldnn_fuse` might compute bad numbers. However, the `conv_bias_mkldnn_fuse` removes the faulty `elementwise_add_mkldnn_op` operator and it works fine again.
Therefore, for the standard use cases, the accuracy with this one and all other resnet50 fuses is the same for both `capi` and `transpiler` unless conv-bias fuse is not present. Link to `conv_bias_mkldnn_fuse` pull request: https://github.com/PaddlePaddle/Paddle/pull/13368.
Do you have a test case for `conv-elementwise_add-batch_norm` which I could use to test the other fuse with it?

**Comment by luotao1** [comment]
> Do you have a test case for conv-elementwise_add-batch_norm which I could use to test the other fuse with it

You can use `ctest -R test_analyzer_resnet50 -V` with the latest develop code

**Comment by sfraczek** [comment]
I used it for conv+bn and it works but not conv-eltwise_add(bias)-bn. Do you have anything else?

**Comment by luotao1** [comment]
The model in `test_analyzer_resnet50` is `conv+elementwise_add+batch_norm`.

**Comment by sfraczek** [comment]
I see you have added this test which is different from what I was testing it on https://github.com/PaddlePaddle/Paddle/pull/13495

**Comment by luotao1** [comment]
You can use the resnet50 model in develop branch, this model is used online.

**Comment by luotao1** [comment]
You can merge with the latest code to see whether `conv+bn` fuse pass is successful in `test_analyzer_resnet50` unit test.

**Comment by sfraczek** [comment]
I tried to make target test_analyzer_resnet50 but there is no such target. I ran `make test_analyzer_resnet50 -j 9`. Can you help?

**Comment by sfraczek** [comment]
@luotao1 You were right. I did as you said and it passed all three tests detecting 53 subgraphs.

**Comment by luotao1** [comment]
@sfraczek I test this PR with #13776, but `test_analyzer_resnet50` fails on `WITH_MKL`.

**Comment by luotao1** [comment]
if `WITH_FAST_MATH=OFF`, this PR is successful in `test_analyzer_resnet50`. But if `WITH_FAST_MATH=ON`, it has diff in `Analyzer_resnet50.compare` unit-test in both MKL/MKLDNN environment.

**This PR is very important,** and if it can be merged, we can skip the difference between MKLDNN batch_norm and Eigen batch_norm at first.

**Comment by ghost** [comment]
Baidu trial to disable use_fast_math (PR13754) in Paddle/EIGEN but soon found lots of errors in many models (see issue 13771). So they have to revert it (via PR13776). Now their idea is that BN OP itself will NEVER appear in graph after proper fusion (e.g. conv+bn, conv+eletwise_add+bn). So it doesn't matter MKLDNN use a high precision sqrt (for BN computation) because MKLDNN BN will NOT be called. The sqrt calculation is now implemented in fusion pass. So what we need do now is to use a "EIGEN fast-math similar" sqrt in our conv+bn pass (now std::sqrt is used. See https://github.com/PaddlePaddle/Paddle/pull/13486/files#diff-60cf61881a474bd4944e97c3d2683e31R149)

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+  auto* variance_tensor =
+      scope->FindVar(bn_variance.Name())->GetMutable<LoDTensor>();
+  auto* mean_tensor = scope->FindVar(bn_mean.Name())->GetMutable<LoDTensor>();
+
+  auto std_tensor = LoDTensor();
+  std_tensor.Resize(bn_bias_tensor.dims());
+  std_tensor =
+      tensor_apply(*variance_tensor, [](float x) { return x + 1e-5f; });
+
+  tensor_apply_inplace(&std_tensor, std::sqrt);
```
Comment (luotao1) [inline paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: Could you use the same sqrt function with `EIGEN_USE_FAST`?

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+
+  auto* scale_tensor = scope->FindVar(bn_scale.Name())->GetMutable<LoDTensor>();
+  auto* variance_tensor =
+      scope->FindVar(bn_variance.Name())->GetMutable<LoDTensor>();
+  auto* mean_tensor = scope->FindVar(bn_mean.Name())->GetMutable<LoDTensor>();
+
+  auto std_tensor = LoDTensor();
+  std_tensor.Resize(bn_bias_tensor.dims());
+  std_tensor =
+      tensor_apply(*variance_tensor, [](float x) { return x + 1e-5f; });
```
Comment (luotao1) [inline paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: It's better not use 1e-5f directly.
You should read the value of attribute `epsilon` of batch_norm op.

Comment (sfraczek) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: ok

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc (hunk 12 above)
Comment (sfraczek) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: I have trouble with testing it on develop branch with merged this PR. 
```
test_analyzer_resnet50: /home/sfraczek/source/Paddle/paddle/fluid/string/tinyformat/tinyformat.h:563: const char*
 paddle::string::tinyformat::detail::streamStateFromFormat(std::ostream&, bool&, int&, const 
char*, const paddle::string::tinyformat::detail::FormatArg*, int&, int): Assertion `0 && 
"tinyformat: Not enough conversion specifiers in format string"' failed.
```
I have no idea what this is. I will push it as fast as I can fix it.

Comment (sfraczek) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: `18% tests passed, 328 tests failed out of 399`

Comment (sfraczek) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: I pushed it anyway without testing since the same error is in almost every test

**Comment by luotao1** [comment]
@sfraczek please add `test=develop` in your commit to pass the CI

**Comment by sfraczek** [comment]
I'm rewriting all tensor operations to Eigen.

**Review by luotao1 — APPROVED** [review]
LGTM. You can rewrite in next PR.

**Comment by sfraczek** [comment]
ok thanks

**Pull request merged**